### PR TITLE
BUG : Generate static apparearance for fields

### LINKED ISSUE #1306: update_page_form_field_values prevents barcode (pdf417) from displaying correctly
Here's a description of the issue
https://stackoverflow.com/questions/73533672/pypdf2-doesnt-display-barcode-pdf417-after-using-update-page-form-field-value/73535421#73535421

[pdf417](https://pypi.org/project/pdf417/) barcode isn't displayed after using update_page_form_field_values()

## Environment

Which environment were you using when you encountered the problem?

* Windows 10
* Visual Studio Code
* python 3.10.6

```bash
$ python -m platform
Windows-10-10.0.19043-SP0

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.3
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from PyPDF2 import PdfReader, PdfWriter


reader = PdfReader("i-130.pdf", strict=True)
pager = reader.pages[0]
field_dict = {"Pt2Line4b_GivenName[0]": "Mark"}

writer = PdfWriter(strict=True)
writer.add_page(pager)
writer.update_page_form_field_values(pager, fields=field_dict)

with open("newfile.pdf", "wb") as fh:
    writer.write(fh)
```

The PDF: https://www.uscis.gov/sites/default/files/document/forms/i-130.pdf - you can add it to the tests

## The issue

It works, just doesn't work correctly

### PR DESCRIPTION
**Pull request by pubpub-zz**
Create/Update AP entries to not require Need_appearances flags

closes #1306

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma 
ready for review

File: pypdf/_writer.py [hunk 1]
```diff
@@ -332,7 +330,7 @@ def _add_page(
         pages[NameObject(PA.COUNT)] = NumberObject(page_count + 1)
         return page
 
-    def set_need_appearances_writer(self) -> None:
+    def set_need_appearances_writer(self, state: bool = True) -> None:
```
Comment (MartinThoma) [inline pypdf/_writer.py:333]:
```suggestion
    def set_need_appearances_writer(self, state: bool = True) -> None:
    """
    Sets the "NeedAppearances" flag in the PDF writer.

    The "NeedAppearances" flag indicates whether the appearance dictionary
    for form fields should be automatically generated by the PDF viewer or
    if the embedded appearence should be used.

    Args:
        state: The actual value of the NeedAppearances flag.

    Returns:
        None
    """
```

File: pypdf/_writer.py [hunk 2]
```diff
@@ -801,11 +803,85 @@ def _get_qualified_field_name(self, parent: DictionaryObject) -> Optional[str]:
...
+        if isinstance(dr, IndirectObject):
+            dr = dr.get_object()
+        if fnt in dr:
+            dct[NameObject("/Resources")] = DictionaryObject(
+                {
+                    NameObject("/Font"): DictionaryObject(
+                        {NameObject(fnt): dr[fnt].indirect_reference}
+                    )
+                }
+            )
```
Comment (MartinThoma) [inline pypdf/_writer.py:865]:
```suggestion
        # Calculate rectangle dimensions
        _rct = cast(RectangleObject, field[AA.Rect])
        rct = RectangleObject((0, 0, _rct[2] - _rct[0], _rct[3] - _rct[1]))

        # Extract font information
        font_properties: Any = (
            cast(str, field[AA.DA]).replace("\n", " ").replace("\r", " ").split(" ")
        )
        font_name = font_properties[font_properties.index("Tf") - 2]
        font_height = float(font_properties[font_properties.index("Tf") - 1])
        y_offset = rct.height - 1 - font_height

        # Retrieve field text and selected values
        field_flags = field.get(FA.Ff, 0)
        if field.get(FA.FT, "/Tx") == "/Ch" and field_flags & FA.FfBits.Combo == 0:
            txt = "\n".join(field.get(FA.Opt, {}))
            sel = field.get("/V", [])
            if not isinstance(sel, list):
                sel = [sel]
        else:  # /Tx
            txt = field.get("/V", "")
            sel = []

        # Generate appearance stream
        ap_stream = f"q\n/Tx BMC \nq\n1 1 {rct.width - 1} {rct.height - 1} re\nW\nBT\n{field[AA.DA]}\n".encode()
        for line_number, line in enumerate(txt.replace("\n", "\r").split("\r")):
            if line in sel:
                # may be improved but can not find how get fill working => replaced with lined box
                ap_stream += (
                    f"1 {y_offset - (line_number * font_height * 1.4) - 1} {rct.width - 2} {font_height + 2} re\n"
                    f"0.5 0.5 0.5 rg s\n{field[AA.DA]}\n"
                ).encode()
            if line_number == 0:
                ap_stream += f"2 {y_offset} Td\n".encode()
            else:
                # Td is a relative translation
                ap_stream += f"0 {- font_height * 1.4} Td\n".encode()
            ap_stream += b"(" + str(line).encode("UTF-8") + b") Tj\n"
        ap_stream += b"ET\nQ\nEMC\nQ\n"

        # Create appearance dictionary
        dct = DecodedStreamObject.initialize_from_dictionary(
            {
                NameObject("/Type"): NameObject("/XObject"),
                NameObject("/Subtype"): NameObject("/Form"),
                NameObject("/BBox"): rct,
                "__streamdata__": ByteStringObject(ap_stream),
                "/Length": 0,
            }
        )

        # Retrieve font information from AcroForm dictionary
        dr: Any = cast(
            dict, cast(DictionaryObject, self._root_object["/AcroForm"]).get("/DR", {})
        )
        if isinstance(dr, IndirectObject):
            dr = dr.get_object()
        dr = dr.get("/Font", {})
        if isinstance(dr, IndirectObject):
            dr = dr.get_object()

        # Update Resources with font information if necessary
        if font_name in dr:
            dct[NameObject("/Resources")] = DictionaryObject(
                {
                    NameObject("/Font"): DictionaryObject(
                        {NameObject(font_name): dr[font_name].indirect_reference}
                    )
                }
            )
```

**Comment by MartinThoma** [comment]
I have two suggestions to improve readability. Besides that, it looks good to me :+1: :-)

File: pypdf/_writer.py [hunk 3]
```diff
@@ -332,7 +330,20 @@ def _add_page(
...
+    The "NeedAppearances" flag indicates whether the appearance dictionary
+    for form fields should be automatically generated by the PDF viewer or
+    if the embedded appearence should be used.
+
+    Args:
+        state: The actual value of the NeedAppearances flag.
+
+    Returns:
+        None
+    """
```
Comment (MartinThoma) [inline pypdf/_writer.py:346]:
```suggestion
        """
        Sets the "NeedAppearances" flag in the PDF writer.

        The "NeedAppearances" flag indicates whether the appearance dictionary
        for form fields should be automatically generated by the PDF viewer or
        if the embedded appearence should be used.

        Args:
            state: The actual value of the NeedAppearances flag.

        Returns:
            None
        """
```

File: pypdf/_writer.py [hunk 4]
```diff
@@ -801,11 +816,96 @@ def _get_qualified_field_name(self, parent: DictionaryObject) -> Optional[str]:
...
+            n = field[AA.AP]["/N"].indirect_reference.idnum  # type: ignore
+            self._objects[n - 1] = dct
+            dct.indirect_reference = IndirectObject(n, 0, self)
+
     def update_page_form_field_values(
         self,
         page: PageObject,
         fields: Dict[str, Any],
         flags: FieldFlag = OPTIONAL_READ_WRITE_FIELD,
+        auto_regen: Optional[bool] = True,
```
Comment (MartinThoma) [inline pypdf/_writer.py:908]:
```suggestion
        auto_regenerate: Optional[bool] = True,
```

File: pypdf/_writer.py [hunk 5]
```diff
@@ -821,57 +921,57 @@ def update_page_form_field_values(
             flags: An integer (0 to 7). The first bit sets ReadOnly, the
                 second bit sets Required, the third bit sets NoExport. See
                 PDF Reference Table 8.70 for details.
+            auto_regen: set/unset the need_appearances flag ;
```
Comment (MartinThoma) [inline pypdf/_writer.py:924]:
```suggestion
            auto_regenerate: set/unset the need_appearances flag ;
```

File: pypdf/_writer.py [hunk 6]
```diff
@@ -821,57 +921,57 @@ def update_page_form_field_values(
...
+                the flag is unchanged if auto_regen is None
```
Comment (MartinThoma) [inline pypdf/_writer.py:925]:
```suggestion
                the flag is unchanged if auto_regenerate is None
```

File: pypdf/_writer.py [hunk 7]
```diff
@@ -821,57 +921,57 @@ def update_page_form_field_values(
...
         """
-        self.set_need_appearances_writer()
+        if isinstance(auto_regen, bool):
+            self.set_need_appearances_writer(auto_regen)
```
Comment (MartinThoma) [inline pypdf/_writer.py:928]:
```suggestion
        if isinstance(auto_regenerate, bool):
            self.set_need_appearances_writer(auto_regenerate)
```

File: tests/test_writer.py [hunk 8]
```diff
@@ -1341,3 +1341,63 @@ def test_iss1767():
...
+            "Liste2": ["Lst1", "Lst3"],
+            "DropList1": "DropListe3",
+        },
+        auto_regen=False,
+    )
+    del writer.pages[0]["/Annots"][1].get_object()["/AP"]["/N"]
+    writer.update_page_form_field_values(
+        writer.pages[0],
+        {"Text1": "my Text1", "Text2": "ligne1\nligne2\nligne3"},
+        auto_regen=False,
```
Comment (MartinThoma) [inline tests/test_writer.py:1376]:
```suggestion
        auto_regenerate=False,
    )
    del writer.pages[0]["/Annots"][1].get_object()["/AP"]["/N"]
    writer.update_page_form_field_values(
        writer.pages[0],
        {"Text1": "my Text1", "Text2": "ligne1\nligne2\nligne3"},
        auto_regenerate=False,
```

**Pull request merged**